pkg/api/subscribeWorker.go

```diff
@@ -0,0 +1,196 @@
+	log *zap.Logger,
+	store *sql.DB,
+) (*subscribeWorker, error) {
+	log = log.Named("subscribeWorker")
+	q := queries.New(store)
+	pollableQuery := func(ctx context.Context, lastSeen db.VectorClock, numRows int32) ([]queries.GatewayEnvelope, db.VectorClock, error) {
+		envs, err := q.
+			SelectGatewayEnvelopes(
+				ctx,
+				*db.SetVectorClock(&queries.SelectGatewayEnvelopesParams{}, lastSeen),
+			)
+		if err != nil {
+			return nil, lastSeen, err
+		}
+		for _, env := range envs {
+			// TODO(rich) Handle out-of-order envelopes
+			lastSeen[uint32(env.OriginatorNodeID)] = uint64(env.OriginatorSequenceID)
+		}
+		return envs, lastSeen, nil
+	}
+
+	vc, err := q.SelectVectorClock(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	subscription := db.NewDBSubscription(
+		ctx,
+		log,
+		pollableQuery,
+		db.ToVectorClock(vc),
+		db.PollingOptions{
+			Interval: SubscribeWorkerPollTime,
+			NumRows:  subscribeWorkerPollRows,
+		},
+	)
+	dbChan, err := subscription.Start()
+	if err != nil {
+		return nil, err
+	}
+	worker := &subscribeWorker{
+		ctx:                 ctx,
+		log:                 log,
+		dbSubscription:      dbChan,
+		globalListeners:     make([]listener, 0),
+		originatorListeners: make(map[uint32][]listener),
+		topicListeners:      make(map[string][]listener),
+	}
+
+	go worker.start()
+
+	return worker, nil
+}
+
+func (s *subscribeWorker) start() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case new_batch := <-s.dbSubscription:
+			for _, row := range new_batch {
+				s.dispatch(&row)
+			}
+		}
+	}
+}
+
+func (s *subscribeWorker) dispatch(
+	row *queries.GatewayEnvelope,
+) {
+	bytes := row.OriginatorEnvelope
+	env := &message_api.OriginatorEnvelope{}
+	err := proto.Unmarshal(bytes, env)
+	if err != nil {
+		s.log.Error("Failed to unmarshal envelope", zap.Error(err))
+		return
+	}
+	for _, listener := range s.originatorListeners[uint32(row.OriginatorNodeID)] {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
+		default: // TODO(rich) Close and clean up channel
+		}
+	}
+	for _, listener := range s.topicListeners[hex.EncodeToString(row.Topic)] {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
+		default:
+		}
+	}
+	for _, listener := range s.globalListeners {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
+		default:
+		}
+	}
+}
+
+func (s *subscribeWorker) addListeners(
+	requests []*message_api.BatchSubscribeEnvelopesRequest_SubscribeEnvelopesRequest,
+) (<-chan []*message_api.OriginatorEnvelope, error) {
+	subscribeAll := false
+	topics := make(map[string]bool, len(requests))
+	originators := make(map[uint32]bool, len(requests))
+
+	if len(requests) > maxSubscriptionsPerClient {
+		// When a client subscribes to too many originators or topics, we treat it as a request to
+		// subscribe to all instead of throwing an error. We rely on the client's existing
+		// filtering logic rather than forcing clients to respond to an error.
```

Comment:
I believe this is different from the old implementation. It means that we can set `maxSubscriptionsPerClient` lower and arbitrarily modify it, and use less memory/iterate through fewer subscriptions.

I suspect this is a better behavior for clients to handle, because clients need to have filtering logic anyway. It does mean that clients have to have the ability to discard payloads that are not relevant to them.

Reply:
Feels a little odd to change the meaning of the request silently. At the very least we should log it, but it might be more intuitive to just return an error and have the caller resubscribe to all. Or, since we control the client, just have it be smart enough to create the right subscription based on the limit the first time.

Reply:
I'm just thinking through the client logic, and think it might be simpler this way?

Option 1: List in the docs that at an unspecified number of batched requests, the stream will subscribe to all updates. Clients should always be filtering the stream by topic regardless (esp if they subscribe to multiple topics), the only addition is they should drop any irrelevant payloads.

Option 2: List in the docs a specific max number of batched requests. Clients need to track the number of batched requests they issue, and switch to a subscribeAll subscription at that number (including if they add requests via the bidirectional stream, they would need to tear it down and re-establish). Once at that number, they need to implement the logic to filter the stream and drop irrelevant payloads regardless.

WDYT? I don't have a strong opinion, so happy to switch to an error if you still feel it is unintuitive.

Reply:
Are we imagining that this behaviour is going to be extended to clients? Like at some point my phone might get the firehose of all messages on XMTP? A public figure user is going to subscribe to thousands of topics since they have thousands of conversations...but that's still going to be a very different throughput to getting every message on XMTP since most of their subscriptions would be dormant.

Reply:
For topic subscriptions, we should be expecting that users can be subscribed to tens of thousands of topics on the client. We can maybe adjust the `subscribe_all_messages` behaviour on the client to only subscribe to consented conversations once we have consent native to V3, which'll lower the number to something reasonable. But still could be thousands.

Bots could have hundreds of thousands of convos and might at least have the compute to handle the firehose. But it's a lot more outbound bandwidth for us to stream all messages than it is to listen to 100k topics that are mostly idle.

Reply:
After our chat, the thing I misunderstood is that the client will always respond to an error by subscribing to all, which is not true - a client may choose to subscribe to fewer topics, or split them across multiple subscriptions. Changed to an error!